src/mscorlib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,365 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Globalization;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// A helper class which allows interacting with dynamically loaded modules.
+    /// </summary>
+    /// <remarks>
+    /// This is similar in principle to <see cref="DllImportAttribute"/>, but it can be used where
+    /// the module name or export names are not known at compile time.
+    /// </remarks>
+    public sealed partial class NativeLibrary
+    {
+        // Per PInvokeStaticSigInfo::PInvokeStaticSigInfo, the default character set is ANSI on all platforms.
+        private const CharSet FallbackCharSet = CharSet.Ansi;
+
+        // The CLR's internal HINSTANCE, which may not match the OS-provided handle.
+        private readonly IntPtr _hInstance;
+
+        private NativeLibrary(IntPtr hModule)
+        {
+            if (!IsValidModuleHandle(hModule))
+            {
+                throw new ArgumentException(
+                    paramName: nameof(hModule),
+                    message: SR.Argument_InvalidHandle);
+            }
+
+            _hInstance = hModule;
+        }
+
+        /// <summary>
+        /// The raw OS-provided handle for this module.
+        /// </summary>
+        /// <remarks>
+        /// The caller can query this handle for information about the loaded module
+        /// but should not attempt to free or otherwise manipulate this handle.
+        /// </remarks>
+        public IntPtr Handle => OperatingSystemHandle;
+
+        // This is a managed equivalent to NDirectMethodDesc::FindEntryPoint.
+        private TDelegate CreateDelegateCore<TDelegate>(string name, bool exactSpelling, RuntimeMethodInfo delegateInvokeMethod) where TDelegate : class
+        {
+            Debug.Assert(typeof(TDelegate) is RuntimeType);
+            Debug.Assert(((RuntimeType)typeof(TDelegate)).IsDelegate());
+            Debug.Assert(name != null);
+            Debug.Assert(delegateInvokeMethod != null);
+
+            // If we allow lookup by ordinal value and an ordinal is specified ("#1234", where
+            // the value immediately following the '#' can fit into a WORD), then process the
+            // ordinal and skip all other logic in this routine.
+
+            if (AllowLocatingFunctionsByOrdinal && name.Length >= 1 && name[0] == '#')
```

Comment:
Is this functionality we need to support?  It's unfortunate that a new API whose purpose seems targeted at being cross-platform would have platform-specific behavioral differences designed in.